execution-engine/blessed-contracts/pos/src/stakes.rs

```diff
@@ -55,6 +60,7 @@ impl Stakes {
             .collect();
         // Remove and add urefs to update the contract's known urefs accordingly.
         for (name, _) in contract_api::list_known_urefs() {
+            // Mateusz: Why are we removing `name` from the `new_urefs` set?
```

Comment:
So that we don't cause a redundant write operation below, if it's already there.

Reply:
What if a validator is trying to bond additional amount to their already bonded stake. This will discard new bonding event and keep the old one. In that case we should be adding up these values.

Reply:
NVM we do that in `Stakes::bond` method.

Reply:
No, wait. If it's already there with an old value we won't write new one.

Reply:
Yes: The uref _name_ alone encodes both the validator ID and the amount. If the uref name we want to create for a validator already exists, then that means they also already have the correct amount.

Reply:
If it adds new stakes then we won't remove the old entry isn't it? And we end up with two entries for the same validator - old value and new one.

Reply:
OK, if the name is different (new stake) then we remove it. :man_facepalming:

Reply:
I'm not sure I'm following:
Say validator Bob has 5 tokens stakesd, and add 3. Then `list_known_urefs` will contain `"v_bob_5"`, and `new_urefs` will contain `"v_bob_8"`. So `new_urefs(remove("v_bob_5"))` will return `false`, and therefore we will remove that uref.
`new_urefs` still contains `"v_bob_8"`, so we will add that below in line 69.

Reply:
No no, I am not following :)